server/etcdserver/v3_server.go

```diff
@@ -156,8 +168,30 @@ func (s *EtcdServer) DeleteRange(ctx context.Context, r *pb.DeleteRangeRequest)
 	return resp.(*pb.DeleteRangeResponse), nil
 }
 
+// firstCompareKey returns first non-empty key in the list of comparison operations.
+func firstCompareKey(c []*pb.Compare) string {
+	for _, op := range c {
+		key := string(op.GetKey())
+		if key != "" {
+			return key
+		}
+	}
+	return ""
+}
+
 func (s *EtcdServer) Txn(ctx context.Context, r *pb.TxnRequest) (*pb.TxnResponse, error) {
 	readOnly := txn.IsTxnReadonly(r)
+
+	var span trace.Span
+	ctx, span = traceutil.Tracer.Start(ctx, "txn", trace.WithAttributes(
+		attribute.String("first_compare_key", firstCompareKey(r.GetCompare())),
+		attribute.Int("compare_len", len(r.GetCompare())),
+		attribute.Int("success_len", len(r.GetSuccess())),
```

Comment:
Maybe add `attribute.Int("success_first_op", "put")`

Reply:
`success_first_key` is possible, `success_first_op` refers to `oneof` field making it non-trivial to implement

Reply:
Followed up in https://github.com/etcd-io/etcd/pull/20511